# Використання Pandas

In [22]:
import pandas as pd
import numpy as np
import timeit
from datetime import datetime
#Зчитаємо файл
df = pd.read_csv('C:\\Users\\Anastasiia\\Desktop\\household_power_consumption.txt', sep = ';',  decimal='/')
df.head()

Date      Time Global_active_power Global_reactive_power  Voltage   
0  16/12/2006  17:24:00               4.216                 0.418  234.840  \
1  16/12/2006  17:25:00               5.360                 0.436  233.630   
2  16/12/2006  17:26:00               5.374                 0.498  233.290   
3  16/12/2006  17:27:00               5.388                 0.502  233.740   
4  16/12/2006  17:28:00               3.666                 0.528  235.680   

  Global_intensity Sub_metering_1 Sub_metering_2 Sub_metering_3  
0           18.400          0.000          1.000         17.000  
1           23.000          0.000          1.000         16.000  
2           23.000          0.000          2.000         17.000  
3           23.000          0.000          1.000         17.000  
4           15.800          0.000          1.000         17.000

In [66]:
start_time = 0.0
def start_count():
    global start_time
    start_time = time.time()
def finish_count():
    print(time.time()-start_time)

In [62]:
#Видалимо строки де немає даних
data1=df[ ~df.Voltage.str.contains("\?") ]
data = pd.DataFrame(data1)
#Перевіримо чи данні видалені
no_data = data.loc[df['Global_intensity'] == '?']
print(no_data)

Empty DataFrame
Columns: [Date, Time, Global_active_power, Global_reactive_power, Voltage, Global_intensity, Sub_metering_1, Sub_metering_2, Sub_metering_3]
Index: []


In [64]:
#Змінимо типи даних
data['Global_active_power'] = data['Global_active_power'].astype(float)
data['Global_reactive_power'] = data['Global_reactive_power'].astype(float)
data['Voltage'] = data['Voltage'].astype(float)
data['Global_intensity'] = data['Global_intensity'].astype(float)
data['Sub_metering_1'] = data['Sub_metering_1'].astype(float)
data['Sub_metering_2'] = data['Sub_metering_2'].astype(float)
data['Sub_metering_3'] = data['Sub_metering_3'].astype(float)

In [80]:
#Знайдемо господарства, де загальна активна споживана потужність перевищує 5 кВт.
start_count()
greater_than_5 = data.loc[(data['Global_active_power'] > 5)]
finish_count()
greater_than_5.head()

0.004009723663330078


Date      Time  Global_active_power  Global_reactive_power  Voltage   
1   16/12/2006  17:25:00                5.360                  0.436   233.63  \
2   16/12/2006  17:26:00                5.374                  0.498   233.29   
3   16/12/2006  17:27:00                5.388                  0.502   233.74   
11  16/12/2006  17:35:00                5.412                  0.470   232.78   
12  16/12/2006  17:36:00                5.224                  0.478   232.99   

    Global_intensity  Sub_metering_1  Sub_metering_2  Sub_metering_3  
1               23.0             0.0             1.0            16.0  
2               23.0             0.0             2.0            17.0  
3               23.0             0.0             1.0            17.0  
11              23.2             0.0             1.0            17.0  
12              22.4             0.0             1.0            16.0

In [81]:
#Знайдемо господарства, де вольтаж перевищує 235 Вольт.
start_count()
greater_than_235 = data.loc[(data['Voltage'] > 235)]
finish_count()
greater_than_235.head()

0.0616917610168457


Date      Time  Global_active_power  Global_reactive_power  Voltage   
4   16/12/2006  17:28:00                3.666                  0.528   235.68  \
5   16/12/2006  17:29:00                3.520                  0.522   235.02   
6   16/12/2006  17:30:00                3.702                  0.520   235.09   
7   16/12/2006  17:31:00                3.700                  0.520   235.22   
14  16/12/2006  17:38:00                4.054                  0.422   235.24   

    Global_intensity  Sub_metering_1  Sub_metering_2  Sub_metering_3  
4               15.8             0.0             1.0            17.0  
5               15.0             0.0             2.0            17.0  
6               15.8             0.0             1.0            17.0  
7               15.8             0.0             1.0            17.0  
14              17.6             0.0             1.0            17.0

In [82]:
#Знайдемо господарства, де сила струму в діапазоні 19-20 Ампер.
start_count()
in_19_20 = data.loc[(data['Global_intensity']>19)&(data['Global_intensity']<20)&(data['Sub_metering_2']>data['Sub_metering_3'])]
finish_count()
in_19_20.head()

0.007335662841796875


Date      Time  Global_active_power  Global_reactive_power   
460  17/12/2006  01:04:00                4.582                  0.258  \
464  17/12/2006  01:08:00                4.618                  0.104   
475  17/12/2006  01:19:00                4.636                  0.140   
476  17/12/2006  01:20:00                4.634                  0.152   
477  17/12/2006  01:21:00                4.652                  0.142   

     Voltage  Global_intensity  Sub_metering_1  Sub_metering_2  Sub_metering_3  
460   238.08              19.6             0.0            13.0             0.0  
464   239.61              19.6             0.0            27.0             0.0  
475   237.37              19.4             0.0            36.0             0.0  
476   237.17              19.4             0.0            35.0             0.0  
477   237.92              19.4             0.0            36.0             0.0

In [83]:
#Виберемо 50000 випадкових значень
start_count()
random_50000 = data.sample(n=50000)
finish_count()
random_50000.head()

0.04011416435241699


Date      Time  Global_active_power  Global_reactive_power   
1355978  15/7/2009  09:02:00                1.538                  0.168  \
100562   24/2/2007  13:26:00                0.216                  0.000   
1284105  26/5/2009  11:09:00                1.466                  0.062   
1711099  18/3/2010  23:43:00                0.424                  0.000   
671791   27/3/2008  05:55:00                0.220                  0.000   

         Voltage  Global_intensity  Sub_metering_1  Sub_metering_2   
1355978   241.39               6.4             2.0             0.0  \
100562    241.31               0.8             0.0             0.0   
1284105   227.71               6.4             0.0             0.0   
1711099   242.49               1.8             0.0             0.0   
671791    244.13               0.8             0.0             0.0   

         Sub_metering_3  
1355978            19.0  
100562              0.0  
1284105            17.0  
1711099             1.0  
671791              0.0

In [84]:
#Знайдемо середнє значення
start_count()
print(random_50000['Sub_metering_1'].mean())
print(random_50000['Sub_metering_2'].mean())
print(random_50000['Sub_metering_3'].mean())
print("***")
finish_count()

1.13988
1.29488
6.53434
***
0.0010006427764892578


In [85]:
#Знайдемо домогосподарства, де після 18:00 Sub_metering_2 має більше знначення 
start_count()
later_than_18 = data.loc[(data['Time'] > '18:00:00')&(data['Sub_metering_2']>data['Sub_metering_3'])&(data['Sub_metering_2']>data['Sub_metering_1'])]
#Розіб'ємо на половини та виведемо для першої половини кожен третій результат, я для другої - кожний четвертий.
first_half = (later_than_18.iloc[:38825,:]).iloc[::3, :]
second_half = (later_than_18.iloc[38825:,:]).iloc[::4, :]
#Поєднаємо їх
result = pd.concat([first_half, second_half])
finish_count()
result.head()

0.05932760238647461


Date      Time  Global_active_power  Global_reactive_power   
40   16/12/2006  18:04:00                4.928                  0.202  \
43   16/12/2006  18:07:00                6.474                  0.144   
46   16/12/2006  18:10:00                3.396                  0.148   
71   16/12/2006  18:35:00                6.072                  0.000   
394  16/12/2006  23:58:00                2.764                  0.066   

     Voltage  Global_intensity  Sub_metering_1  Sub_metering_2  Sub_metering_3  
40    235.01              21.0             0.0            37.0            16.0  
43    231.85              27.8             0.0            37.0            16.0  
46    236.20              15.0             0.0            22.0            18.0  
71    232.48              26.4             0.0            27.0            17.0  
394   241.27              11.4             0.0             1.0             0.0

# Використання Numpy

In [87]:
#Прочитаємо дані як масив numpy
dn = data.to_numpy()
#Знайдемо господарства, де загальна активна споживана потужність перевищує 5 кВт.
start_count()
d1 = dn[np.where(dn[...,2]>5)]
finish_count()
print(d1)

0.03000330924987793
[['16/12/2006' '17:25:00' 5.36 ... 0.0 1.0 16.0]
 ['16/12/2006' '17:26:00' 5.374 ... 0.0 2.0 17.0]
 ['16/12/2006' '17:27:00' 5.388 ... 0.0 1.0 17.0]
 ...
 ['24/11/2010' '07:50:00' 5.172 ... 0.0 38.0 17.0]
 ['24/11/2010' '07:51:00' 5.75 ... 0.0 39.0 17.0]
 ['25/11/2010' '07:21:00' 5.074 ... 1.0 2.0 18.0]]


In [88]:
#Знайдемо господарства, де вольтаж перевищує 235 Вольт.
start_count()
d2 = dn[np.where(dn[...,4]>235)]
finish_count()
print(d2)

0.0807197093963623
[['16/12/2006' '17:28:00' 3.666 ... 0.0 1.0 17.0]
 ['16/12/2006' '17:29:00' 3.52 ... 0.0 2.0 17.0]
 ['16/12/2006' '17:30:00' 3.702 ... 0.0 1.0 17.0]
 ...
 ['26/11/2010' '21:00:00' 0.938 ... 0.0 0.0 0.0]
 ['26/11/2010' '21:01:00' 0.934 ... 0.0 0.0 0.0]
 ['26/11/2010' '21:02:00' 0.932 ... 0.0 0.0 0.0]]


In [89]:
#Знайдемо господарства, де сила струму в діапазоні 19-20 Ампер.
start_count()
d3 = dn[np.where((dn[...,5]>19)&(dn[...,5]<20)&(dn[...,7]>dn[...,8]))]
finish_count()
print(d3)

0.07647871971130371
[['17/12/2006' '01:04:00' 4.582 ... 0.0 13.0 0.0]
 ['17/12/2006' '01:08:00' 4.618 ... 0.0 27.0 0.0]
 ['17/12/2006' '01:19:00' 4.636 ... 0.0 36.0 0.0]
 ...
 ['24/11/2010' '07:54:00' 4.694 ... 0.0 39.0 18.0]
 ['24/11/2010' '07:55:00' 4.602 ... 0.0 40.0 17.0]
 ['24/11/2010' '07:57:00' 4.626 ... 0.0 39.0 17.0]]


In [76]:
#Виберемо 50000 випадкових значень
start_count()
print(dn[np.random.choice(dn.shape[0], 50000, replace = False)])
finish_count()

[['27/4/2007' '03:20:00' 0.202 ... 0.0 0.0 0.0]
 ['13/5/2010' '00:13:00' 0.326 ... 0.0 0.0 0.0]
 ['31/1/2008' '05:25:00' 0.422 ... 0.0 0.0 0.0]
 ...
 ['24/7/2007' '04:12:00' 0.216 ... 0.0 0.0 0.0]
 ['25/11/2007' '17:00:00' 0.298 ... 0.0 0.0 0.0]
 ['7/4/2008' '15:03:00' 0.506 ... 0.0 1.0 0.0]]
0.049845218658447266


In [78]:
#Знайдемо середнє значення
start_count()
print(np.mean(dn[...,6]))
print(np.mean(dn[...,7]))
print(np.mean(dn[...,8]))
print("***")
finish_count()

1.1219233096502186
1.2985199679887571
6.45844735712055
***
0.04952836036682129


In [79]:
#Розіб'ємо на половини та виведемо для першої половини кожен третій результат, я для другої - кожний четвертий.
start_count()
d6 = dn[np.where((dn[...,1]>'18:00:00')&(dn[...,2]>6)&(dn[...,7]>dn[...,6])&(dn[...,7]>dn[...,8]))]
new = np.split(d6, [529, 530])
print(new[0][::3])
print(new[1][::4])
finish_count()

[['16/12/2006' '18:05:00' 6.052 ... 0.0 37.0 17.0]
 ['16/12/2006' '18:08:00' 6.308 ... 0.0 36.0 17.0]
 ['28/12/2006' '20:58:00' 6.386 ... 1.0 36.0 17.0]
 ...
 ['23/3/2008' '20:36:00' 6.708 ... 0.0 73.0 16.0]
 ['20/4/2008' '21:00:00' 8.268 ... 1.0 72.0 17.0]
 ['20/4/2008' '21:03:00' 7.798 ... 1.0 71.0 17.0]]
[['20/4/2008' '21:04:00' 7.756 0.348 232.11 33.4 1.0 72.0 17.0]]
0.08147811889648438


I. Пошук домогосподарств, де загальна активна споживана потужність перевищує 5 кВт, зайняв 0.004009723663330078 с в pandas та 0.03000330924987793 с в numpy, отже, з цю задачу pandas виконав швидше.
II. Пошук домогосподарств, де вольтаж перевищує 235 Вольт, зайняв 0.0616917610168457 с в pandas та 0.0807197093963623 с в numpy, отже, з цю задачу pandas виконав швидше.
III. Пошук домогосподарств, де сила струму в діапазоні 19-20 Ампер та Sub_metering_2 > Sub_metering_3, зайняв 0.007335662841796875 с в pandas та 0.07647871971130371 с в numpy, отже, з цю задачу pandas виконав трохи швидше.
IV. Знаходження середнього значення в pandas зайняло 0.0010006427764892578 с, а в numpy 0.04952836036682129, отже, з цю задачу pandas теж виконав швидше.
V. Знаходження домогосподарств, де після 18:00 Sub_metering_2 має більше знначення та виведення кожного третього та кожного четвертого зачення в першій та другій половинах відповідно зайняло 0.05932760238647461 с в pandas та 0.08147811889648438 с в numpy, отже, pandas виконав цю задачу швидше. 
Для мене особисто pandas зручніший, ніж numpy, і його я можу оцінити на 5/5, тоді як numpy - на 4/5.